src/queue_pair.rs

```diff
@@ -1,551 +1,581 @@
+
+impl QueuePairOp for QPSend<'_> {
+    type Output = ();
+
+    fn submit(&self, qp: &QueuePair, wr_id: WorkRequestId) -> io::Result<()> {
+        qp.submit_send(&self.lms, wr_id)
+    }
+
+    fn should_resubmit(&self, e: &io::Error) -> bool {
+        matches!(e.kind(), io::ErrorKind::OutOfMemory)
+    }
+
+    fn result(&self, wc: WorkCompletion) -> Result<Self::Output, WCError> {
+        wc.result().map(|sz| assert_eq!(sz, self.len))
+    }
+}
+
+/// Queue pair receive operation
+#[derive(Debug)]
+pub struct QPRecv<'lm> {
+    /// the local memory regions
+    lms: Vec<&'lm LocalMemoryRegion>,
+}
+
+impl<'lm> QPRecv<'lm> {
+    /// create a new queue pair receive operation
+    fn new(lms: &[&'lm LocalMemoryRegion]) -> Self {
+        Self { lms: lms.to_vec() }
+    }
+}
+
+impl QueuePairOp for QPRecv<'_> {
+    type Output = usize;
+
+    fn submit(&self, qp: &QueuePair, wr_id: WorkRequestId) -> io::Result<()> {
+        qp.submit_receive(&self.lms, wr_id)
+    }
+
+    fn should_resubmit(&self, e: &io::Error) -> bool {
+        matches!(e.kind(), io::ErrorKind::OutOfMemory)
+    }
+
+    fn result(&self, wc: WorkCompletion) -> Result<Self::Output, WCError> {
+        wc.result()
+    }
+}
+
+/// Queue pair operation state
+#[derive(Debug)]
+enum QueuePairOpsState {
+    /// It's in init state, not yet submitted
+    Init,
+    /// Submit
+    Submit(WorkRequestId, Option<mpsc::Receiver<WorkCompletion>>),
+    /// Sleep and prepare to resubmit
+    PendingToResubmit(
+        Pin<Box<Sleep>>,
+        WorkRequestId,
+        Option<mpsc::Receiver<WorkCompletion>>,
+    ),
+    /// have submitted
+    Submitted(mpsc::Receiver<WorkCompletion>),
+}
+
+/// Queue pair operation wrapper
+#[derive(Debug)]
+pub struct QueuePairOps<Op: QueuePairOp + Unpin> {
+    /// the internal queue pair
+    qp: Arc<QueuePair>,
+    /// operation state
+    state: QueuePairOpsState,
+    /// the operation
+    op: Op,
+}
+
+impl<Op: QueuePairOp + Unpin> QueuePairOps<Op> {
+    /// Create a new queue pair operation wrapper
+    fn new(qp: Arc<QueuePair>, op: Op) -> Self {
+        Self {
+            qp,
+            state: QueuePairOpsState::Init,
+            op,
+        }
+    }
+}
+
+impl<Op: QueuePairOp + Unpin> Future for QueuePairOps<Op> {
+    type Output = io::Result<Op::Output>;
+
+    #[allow(clippy::restriction)]
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let s = self.get_mut();
+        match s.state {
+            QueuePairOpsState::Init => {
+                let (wr_id, recv) = s.qp.event_listener.register();
+                s.state = QueuePairOpsState::Submit(wr_id, Some(recv));
+                Pin::new(s).poll(cx)
+            }
+            QueuePairOpsState::Submit(wr_id, ref mut recv) => {
+                if let Err(e) = s.op.submit(&s.qp, wr_id) {
+                    if s.op.should_resubmit(&e) {
+                        let sleep = Box::pin(sleep(Duration::from_micros(10)));
```

Comment:
Put the delay number to a const.

Reply:
https://github.com/datenlord/async-rdma/pull/28/commits/d59d1679f0aec96edfdfe5735abb0108a2ac7dd3 fixed